src/components/Microsoft.DotNet.UpgradeAssistant.MSBuild/PackageLoader.cs

```diff
@@ -96,57 +109,64 @@ public async Task<IEnumerable<NuGetReference>> GetNewerVersionsAsync(NuGetRefere
+            if (searchResults.Count == 0)
+            {
+                return Enumerable.Empty<NuGetReference>();
+            }
+
+            var results = searchResults
+                .Where(r => currentVersion is null ? true : r.Identity.Version > currentVersion);
 
-            // Query each package source for listed versions of the given package name
-            foreach (var source in packageSources?.Select(p => new PackageSource(p)) ?? _packageSources)
+            if (latestMinorAndBuildOnly)
             {
-                try
-                {
-                    var metadata = await GetSourceRepository(source).GetResourceAsync<PackageMetadataResource>(token).ConfigureAwait(false);
-                    var searchResults = await CallWithRetryAsync(() => metadata.GetMetadataAsync(packageName, includePrerelease: includePreRelease, includeUnlisted: false, _cache, _nugetLogger, token)).ConfigureAwait(false);
-                    var highestVersionResult = searchResults.Select(r => r.Identity.Version).Max(v => v);
-                    if (highestVersionResult is null)
-                    {
-                        continue;
-                    }
+                var max = results
+                    .GroupBy(r => r.Identity.Version)
+                    .Max(t => t.Key);
 
-                    if (highestVersion is null || highestVersionResult > highestVersion)
-                    {
-                        highestVersion = highestVersionResult;
-                    }
-                }
-                catch (NuGetProtocolException)
-                {
-                    _logger.LogWarning("Failed to get package versions from source {PackageSource} due to a NuGet protocol error", source.Source);
-                }
-                catch (HttpRequestException exc)
+                if (results is null)
                 {
-                    _logger.LogWarning("Failed to get package versions from source {PackageSource} due to an HTTP error ({StatusCode})", source.Source, exc.StatusCode);
+                    return Enumerable.Empty<NuGetReference>();
                 }
             }
 
-            if (highestVersion is null)
+            var groupResult = results
+                .OrderByDescending(r => r.Identity.Version)
+                .GroupBy(r =>
+                {
+                    var unsupported = tfms;
+
+                    foreach (var dep in r.DependencySets)
+                    {
+                        foreach (var t in unsupported)
+                        {
+                            if (DefaultCompatibilityProvider.Instance.IsCompatible(dep.TargetFramework, t))
+                            {
+                                unsupported = unsupported.Remove(t);
+                            }
+                        }
+                    }
+
+                    return unsupported.Count;
+                })
+                .OrderBy(g => g.Key)
+                .FirstOrDefault();
```

Comment:
If I'm understanding this right, it appears to be grouping the versions according to how many of the TFMs they don't support and then choosing the one with the least unsupported TFMs. Would it be better to not group and, instead, filter to only versions that have 0 unsupported TFMs?

Reply:
I was wondering something similar. if there's 1 unsupported TFM is it a valid package upgrade?

Reply:
are there scenarios that were built for that would make good unit tests?

Reply:
I had wondered about that and was developing it for the solution where we could do a "best fit". For now, though, I'd rather just replace it with ones that will actually work and we can deal with it later if it doesn't.